docs: add initial 3.5 release notes (#2202)
canonical/charmcraft: docs/release-notes/charmcraft-3.5.rst

Charmcraft 3.5 release notes
============================

> **UNRELEASED**. Expected release date: 7 April 2025

Reviewer: As-is, this will build on `latest`. I don't think we want that.

It would be easier to de-index the page and add it to the blocklist.

Learn about the new features, changes, and fixes introduced in Charmcraft 3.5.

Requirements and compatibility
------------------------------

For development and testing, Charmcraft requires a host with a minimum of 4GB RAM
running a Linux distribution compatible with systemd.

All versions of Charmcraft require the following software:

- systemd
- `snapd`_
- Either `LXD`_ or `Multipass`_

We recommend you install the `Charmcraft snap <https://snapcraft.io/charmcraft>`_. It
comes comes bundled with all its dependencies.

Non-snap installations of Charmcraft have the following dependencies:

- Python 3.10 or higher
- libgit2 1.7
- `skopeo`_
- `Spread`_

What's new
----------

Charmcraft 3.5 brings the following features, integrations, and improvements.

Updated ``init`` templates
~~~~~~~~~~~~~~~~~~~~~~~~~~

The templates generated by ``charmcraft init`` now use the new
:doc:`/reference/platforms` format.

> JSON Schema

Reviewer: Since this change is only (I think) about the schema's addition to the store, the title should be more along the lines of "Added Charmcraft schema to the JSON Schema Store".

Reviewer: I'd still like us to reconsider this heading, as it doesn't fit the model.

~~~~~~~~~~~

> The `JSON Schema Store <https://schemastore.org>`_ now includes a schema for the :ref:`charmcraft-yaml-file`. If your text editor supports getting JSON schemas from

Reviewer: Are we also updating the schema we ship in the package? IDEs support loading schemas as a local file.

> The `JSON Schema Store <https://schemastore.org>`_ now includes a schema for the :ref:`charmcraft-yaml-file`. If your text editor supports getting JSON schemas from
> the schema store, they may already be visible in your text editor.

Reviewer: ```suggestion
The `JSON Schema Store <https://schemastore.org>`_ now includes a schema for the
:ref:`charmcraft-yaml-file`.

If your IDE supports fetching JSON schemas from the store, you can now load the latest.
```

.. image:: /_static/assets/json-schema-example-light.webp
   :width: 451
   :alt: A charmcraft.yaml file open in kate editor showing the mouseover hint for the parts key.

If your editor doesn't support JSON schemas directly but does support the
`Language Server Protocol <https://microsoft.github.io/language-server-protocol/>`_,
you can use a `YAML Language Server <https://snapcraft.io/yaml-language-server>`_

> to provide the feature.

Reviewer: > doesn't support JSON schemas directly

I'm not sure what this means, or whether it's technically correct. Any tool with autocomplete has the rough notion of a language server (a separate module that provides intelligent language parsing), even if they don't use that specific modelling.

Ultimately I don't think it's a distinction worth making in this note. I think we're pushing too far from the topic at hand.

If it's a question of "does my IDE support JSON schemas?", let's leave it to the reader, as you did in the first paragraph. The title of the note will signal to them a topic they may or may not understand, with a feature they may or may not use.

> Improved documentation
> ~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
Documentation improvements
~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Consistent with past notes.

Many people have made countless improvements to Charmcraft's documentation this release

> cycle, from hard to notice things like

Reviewer: I'd drop "cycle".

> cycle, from hard to notice things like

Reviewer: ```suggestion
cycle, from small changes like
```

`using consistent terminology <https://github.com/canonical/charmcraft/pull/2169>`_ to
`major rewrites of tutorials <https://github.com/canonical/charmcraft/pull/2085>`_.

Feature deprecations
--------------------

The following features are deprecated in Charmcraft 3.5 and will be removed in
Charmcraft 4.0.

Windows support
~~~~~~~~~~~~~~~

Charmcraft 3.0 deprecated support for Windows, and the upcoming 4.0 major release will

> remove support for it (`#1810 <https://github.com/canonical/charmcraft/issues/1810>`_).

Reviewer: ```suggestion
remove it (`#1810 <https://github.com/canonical/charmcraft/issues/1810>`_).
```

Windows users should begin preparing to migrate their workflows to `Windows Subsystem
for Linux <https://ubuntu.com/desktop/wsl>`_.

CentOS 7 support
~~~~~~~~~~~~~~~~

> CentOS 7 reached its `end of life
> <https://www.redhat.com/en/topics/linux/centos-linux-eol>`_ on 30 June 2024. Charmcraft
> 3 will continue its current support for CentOS 7, but Charmcraft 4.0 won't support
> CentOS 7 (`#1826 <https://github.com/canonical/charmcraft/issues/1826>`_).

Reviewer: ```suggestion
CentOS 7 reached its `end of life
<https://www.redhat.com/en/topics/linux/centos-linux-eol>`_ on 30 June 2024. Charmcraft
3.0 will continue supporting it, but Charmcraft 4.0 will cease support for it
(`#1826 <https://github.com/canonical/charmcraft/issues/1826>`_).
```
